src/rgw/rgw_op.cc

```diff
@@ -5284,6 +5284,62 @@ int RGWCopyObj::verify_permission(optional_yield y)
 					      RGW_PERM_READ)) { 
 	  return -EACCES;
 	}
+
+  // Check for tagging permissions
+  auto tagging_drctv = s->info.env->get("HTTP_X_AMZ_TAGGING_DIRECTIVE");
+  TaggingDirective itag_der = TaggingDirective::NO_TAGGING_DIRECTIVE;
+  if (tagging_drctv) {
+    if (strcmp(tagging_drctv, "COPY") == 0)
+      itag_der = TaggingDirective::COPY_TAGGING_DIRECTIVE;
+    else if (strcmp(tagging_drctv, "REPLACE") == 0)
+      itag_der = TaggingDirective::REPLACE_TAGGING_DIRECTIVE;
+    else{
+      ldpp_dout(this, 0) << "ERROR: Invalid tagging directive " << dendl;
+      s->err.message = "Unknown tagging directive.";
+      return -EINVAL;
+    }
+  } else {
+      itag_der = TaggingDirective::NO_TAGGING_DIRECTIVE;
+    }
+  TaggingPermission tagging_permission;
+
+  if(Effect::Allow == eval_identity_or_session_policies(this,s->iam_user_policies,
+                                                                  s->env,
+                                                                  rgw::IAM::s3PutObjectTagging,
+                                                                  obj_arn)) {
+      tagging_permission |= TaggingPermission::PUT_TAGGING_PERM;
+
+                                                                  }
+
+  if(Effect::Allow == eval_identity_or_session_policies(this,s->iam_user_policies,
+                                                                  s->env,
+                                                                  rgw::IAM::s3GetObjectTagging,
+                                                                  obj_arn)) {
+      tagging_permission |= TaggingPermission::GET_TAGGING_PERM;
+                                                                  }
+
+  op_ret = s->src_object->get_obj_attrs(s->yield, this);
+  if (op_ret == 0) {
+      attrs = s->src_object->get_attrs();
+      auto tags = attrs.find(RGW_ATTR_TAGS);
+      if (tags != attrs.end()) {
+        //tags present on source object
+        if (tagging_permission == TaggingPermission::PUT_TAGGING_PERM) {
+          if (itag_der == TaggingDirective::COPY_TAGGING_DIRECTIVE || itag_der == TaggingDirective::NO_TAGGING_DIRECTIVE)
+            return -EACCES;
```

Comment:
return -EACCES, because this step would need GET_TAGGING_PERM also? Well its not checked here and same approach needed as mentioned in line 5327 comment

Reply:
Added appropriate comments for understanding cases.